File: compiler/src/steps/must_compile_error_step.py
from pipeline.steps import MacroProcessingStep
from core.macro_registry import MacroContext
from dataclasses import replace
from utils.strutil import IndentedStringIO
from utils.logger import default_logger


class MustCompileErrorVerificationStep(MacroProcessingStep):
    """Step that handles must_compile_error nodes - extracts expectations, processes children, and verifies errors."""
    
    def __init__(self):
        super().__init__()
        self.expectations = []  # Store expectations as we find them
        
    def process_node(self, ctx: MacroContext) -> None:
        """Process node using standard tree walking, handling must_compile_error nodes as we encounter them."""
        from core.node import Macro
        
        # Handle current node if it's must_compile_error
        macro = str(ctx.compiler.get_metadata(ctx.node, Macro))
        if macro == "must_compile_error":
            expected_errors = self._extract_expectations_from_node(ctx, ctx.node)
            if expected_errors is not None:
                self.expectations.append({
                    'node': ctx.node,
                    'expected_errors': expected_errors
                })
            # Don't process children - they were already processed by earlier steps
        else:
            # Process children for non-must_compile_error nodes (standard tree walking pattern)
            for child in ctx.node.children:
                child_ctx = replace(ctx, node=child)
                self.process_node(child_ctx)
        
        # At the solution level, verify all collected expectations
        if ctx.node.content == "67lang:solution":
            self._verify_expectations(ctx, self.expectations)
            self.expectations = []  # Reset for next compilation
        
    def _extract_expectations_from_node(self, ctx: MacroContext, node):
        """Extract expected errors from a must_compile_error node."""
        from utils.error_types import ErrorType
        from core.node import Args
        
        args = ctx.compiler.get_metadata(node, Args)
        default_logger.macro(f"must_compile_error with args: '{args}'")
        
        # Parse expected errors from args: "ERROR_TYPE=line ERROR_TYPE2=line2" or "ERROR_TYPE=+offset"
        expected_errors = {}
        if not args.strip():
            ctx.compiler.compile_error(node, "must_compile_error macro requires arguments in format ERROR_TYPE=line or ERROR_TYPE=+offset", ErrorType.INVALID_MACRO)
            return None
            
        # Get the line number of this must_compile_error node
        must_compile_error_line = node.pos.line if node.pos else 0
            
        for pair in args.split():
            if "=" not in pair:
                ctx.compiler.compile_error(node, f"must_compile_error argument must contain '=': {pair}", ErrorType.INVALID_MACRO)
                return None
            error_type, line_str = pair.split("=", 1)
            
            try:
                if line_str.startswith("+"):
                    # Relative line number: +offset from the must_compile_error line
                    offset = int(line_str[1:])
                    line_num = must_compile_error_line + offset
                else:
                    # Absolute line number
                    line_num = int(line_str)
                expected_errors[line_num] = error_type.strip()
            except ValueError:
                ctx.compiler.compile_error(node, f"invalid line number in must_compile_error: {line_str} (use absolute line number or +offset)", ErrorType.INVALID_MACRO)
                return None
        
        return expected_errors
    
    def _verify_expectations(self, ctx: MacroContext, expectations):
        """Verify that expected errors from must_compile_error macros occurred."""
        from utils.error_types import ErrorType
        
        for expectation in expectations:
            node = expectation['node']
            expected_errors = expectation['expected_errors']
            
            # Build a map of actual errors by line
            # Support multiple compile errors from a single line
            actual_errors_by_line = {}
            for error in ctx.compiler.compile_errors:
                line = error["line"]
                error_type = error.get("error_type", "UNKNOWN_ERROR")
                if line not in actual_errors_by_line:
                    actual_errors_by_line[line] = []
                actual_errors_by_line[line].append(error_type)
            
            # Track which errors were consumed by this expectation
            consumed_error_lines = []
            
            # Check each expected error
            for expected_line, expected_type in expected_errors.items():
                if expected_line not in actual_errors_by_line:
                    # Show nearby errors to help with debugging
                    nearby_errors = []
                    for line in sorted(actual_errors_by_line.keys()):
                        if abs(line - expected_line) <= 3:  # Show errors within 3 lines
                            nearby_errors.append(f"line {line}: {actual_errors_by_line[line]}")
                    
                    nearby_info = f" (nearby errors: {'; '.join(nearby_errors)})" if nearby_errors else ""
                    ctx.compiler.compile_error(
                        node, 
                        f"expected {expected_type} error on line {expected_line} but no error found{nearby_info}",
                        ErrorType.ASSERTION_FAILED
                    )
                elif expected_type not in actual_errors_by_line[expected_line]:
                    ctx.compiler.compile_error(
                        node,
                        f"expected {expected_type} error on line {expected_line} but found {actual_errors_by_line[expected_line]}",
                        ErrorType.ASSERTION_FAILED
                    )
                else:
                    # Error matched, mark it for consumption
                    consumed_error_lines.append(expected_line)
            
            # Remove consumed errors from compile_errors
            # For now, remove all errors from consumed lines - this could be improved
            # to only remove the specific error types that were expected
            ctx.compiler.compile_errors = [
                error for error in ctx.compiler.compile_errors 
                if error["line"] not in consumed_error_lines
            ]

File: compiler/src_ts/pipeline/steps/must_compile_error_step.ts
// pipeline/steps/must_compile_error_verification.ts

import { Macro_context, Macro_registry } from "../../core/macro_registry.ts";
import { Args, Macro, Node } from "../../core/node.ts";
import { default_logger } from "../../utils/logger.ts";
import { ErrorType } from "../../utils/error_types.ts";
import { Macrocosm } from "../../core/macrocosm.ts";
import { Macro_error } from "../../core/exceptions.ts";
import { unroll_parent_chain } from "./utils.ts";

// ----------------------- helpers -----------------------

function mv_to_i(v: unknown): number | null {
  if (typeof v === "number") {
    return v | 0;
  }
  if (typeof v === "string") {
    const i = parseInt(v, 10);
    return Number.isNaN(i) ? null : i;
  }
  return null;
}

function mv_to_s(v: unknown): string | null {
  if (typeof v === "string") {
    return v;
  }
  return null;
}

// ----------------------- main step -----------------------

export class Must_compile_error_context implements Macro_context {
  node: Node;
  compiler: Macrocosm;
  registry: Macro_registry<Must_compile_error_context>;
  expectations: Array<Expectation> = [];

  constructor(opts: {
    node: Node;
    compiler: Macrocosm,
    registry: Macro_registry<Must_compile_error_context>,
    expectations: Expectation[],
  }) {
    this.node = opts.node;
    this.compiler = opts.compiler;
    this.registry = opts.registry;
    this.expectations = opts.expectations || [];
  }

  clone_with(opts: { node: Node }): Must_compile_error_context {
    return new Must_compile_error_context({
      node: opts.node ?? this.node,
      compiler: this.compiler,
      registry: this.registry,
      expectations: this.expectations,
    });
  }

  apply(): null {
    const macroName = String(this.compiler.get_metadata(this.node, Macro));

    if (macroName === "must_compile_error") {
      const exp = this.extract_expectations_from_node(this, this.node);
      if (exp) {
        this.expectations.push(new Expectation(this.node, exp));
      }
    } else {
      for (const child of this.node.children) {
        const childCtx = this.clone_with({ node: child });
        childCtx.apply();
      }
    }

    if (this.node.content === "67lang:solution") {
      this.verify_expectations(this, this.expectations);
      this.expectations = [];
    }

    return null;
  }

  // ------------------- extraction -------------------

  private extract_expectations_from_node(
    ctx: Macro_context,
    node: Node,
  ): Map<number, string> | null {
    const args = String(ctx.compiler.get_metadata(node, Args) ?? "");

    if (args.trim() === "") {
      ctx.compiler.error_tracker.fail({
        node,
        message: "must_compile_error macro requires arguments in format ERROR_TYPE=line or ERROR_TYPE=+offset",
        type: ErrorType.INVALID_MACRO,
      });
      return null;
    }

    const baseLine = node.pos?.line ?? 0;
    const expected = new Map<number, string>();

    for (const pair of args.split(/\s+/)) {
      const [rawType, rawNum] = pair.split("=", 2).map((x) => x.trim());

      
      ctx.compiler.error_tracker.assert(
        rawType !== undefined && rawNum !== undefined,
        {
          node,
          message: `must_compile_error argument must contain '=': ${pair}`,
          type: ErrorType.INVALID_MACRO,
        }
      );

      let lineNum: number | null = null;

      if (rawNum.startsWith("+")) {
        const off = rawNum.slice(1);
        const i = parseInt(off, 10);
        lineNum = Number.isNaN(i) ? null : baseLine + i;
      } else {
        const i = parseInt(rawNum, 10);
        lineNum = Number.isNaN(i) ? null : i;
      }

      if (lineNum === null) {
        ctx.compiler.error_tracker.fail({
          node,
          message: `invalid line number in must_compile_error: ${rawNum} (use absolute line number or +offset)`,
          type: ErrorType.INVALID_MACRO,
        });
        return null;
      }

      expected.set(lineNum, rawType);
    }

    return expected;
  }

  // ------------------- verification -------------------

  private verify_expectations(
    ctx: Macro_context,
    expectations: Array<Expectation>,
  ): void {
    function line(err: Macro_error): number {
      return err.node.pos?.line ?? 0;
    }

    const all_ours = ctx.compiler.error_tracker.all.filter((e) => unroll_parent_chain(e.node).includes(ctx.node));

    const actualByLine: [/*line*/ number, ErrorType, Macro_error][] = [];

    for (const err of all_ours) {
      actualByLine.push([line(err), err.type, err]);
    }

    const consumed: Macro_error[] = [];

    for (const exp of expectations) {
      for (const [expectedLine, expectedType] of exp.expected_errors) {
        ctx.compiler.error_tracker.safely(ctx, () => {
          const matched = actualByLine.filter(([line, type]) =>
            line === expectedLine && type === expectedType
          );
          ctx.compiler.error_tracker.assert(
            matched.length > 0,
            {
              node: exp.node,
              message: `expected compile error of type ${expectedType} at line ${expectedLine}, but none found`,
              type: ErrorType.EXPECTATION_VIOLATED,
            }
          );
          for (const [, , err] of matched) {
            consumed.push(err);
          }
        });
      }
    }
    
    for (const error of consumed) {
      ctx.compiler.error_tracker.tolerate(error, "matched by `must_compile_error` macro");
    }
  }
}

class Expectation {
  public node: Node;
  public expected_errors: Map<number, string>;

  constructor(node: Node, expected: Map<number, string>) {
    this.node = node;
    this.expected_errors = expected;
  }
}
